### PR TITLE
add v8::EscapableHandleScope

### PR DESCRIPTION
**Pull request by bartlomieju**

### CODE REVIEW

File: src/handle_scope.rs [hunk 1]
```diff
@@ -55,3 +70,61 @@ impl<'sc> AsMut<Isolate> for HandleScope<'sc> {
...
+  fn drop(&mut self) {
+    unsafe { v8__EscapableHandleScope__DESTRUCT(self) }
+  }
+}
+
+impl<'sc> AsRef<EscapableHandleScope<'sc>> for EscapableHandleScope<'sc> {
+  fn as_ref(&self) -> &Self {
+    self
+  }
+}
```
Comment (ry) [inline src/handle_scope.rs]: what is this for?

File: src/handle_scope.rs [hunk 2]
```diff
@@ -55,3 +70,61 @@ impl<'sc> AsMut<Isolate> for HandleScope<'sc> {
...
+      scope.assume_init()
+    }
+  }
+
+  /// Pushes the value into the previous scope and returns a handle to it.
+  /// Cannot be called twice.
+  pub fn escape<'parent>(
+    &mut self,
+    mut value: Local<'sc, Value>,
+  ) -> Local<'parent, Value> {
```
Comment (ry) [inline src/handle_scope.rs]: I guess this is right. @piscisaureus ?

File: tests/test_api.rs [hunk 3]
```diff
@@ -93,6 +93,63 @@ fn test_string() {
...
+  params.set_array_buffer_allocator(v8::Allocator::new_default_allocator());
+  let mut isolate = v8::Isolate::new(params);
+  let mut locker = v8::Locker::new(&isolate);
+  isolate.enter();
+  v8::HandleScope::enter(&mut locker, |scope1| {
+    // After dropping EscapableHandleScope, we should be able to 
+    // read escaped values.  
+
+    let mut escapable_scope = v8::EscapableHandleScope::new(scope1);
+    let mut scope: v8::HandleScope = unsafe { std::mem::transmute_copy(&escapable_scope) };
```
Comment (ry) [inline tests/test_api.rs]: :/

File: tests/test_api.rs [hunk 4]
```diff
@@ -93,6 +93,63 @@ fn test_string() {
...
+    let l2: Local<v8::Integer> = cast(escapable_scope.escape(l2));
+    drop(escapable_scope);
+    assert_eq!(l2.value(), 456);
+
+    let mut escapable_scope = v8::EscapableHandleScope::new(scope1);
+    let mut scope: v8::HandleScope = unsafe { std::mem::transmute_copy(&escapable_scope) };
+    let l3: Local<v8::Value> = cast(v8::Number::new(&mut scope, 78.9));
+    let l3: Local<v8::Number> = cast(escapable_scope.escape(l3));
+    drop(escapable_scope);
+    assert_eq!(l3.value(), 78.9);
```
Comment (ry) [inline tests/test_api.rs]: There are 3 v8::Number::new tests - maybe there can be just one

File: tests/test_api.rs (hunk 3 above)
Comment (bartlomieju) [reply tests/test_api.rs]: Worst thing is this is very bad... `HandleScope` has size 3 and `EscapabaleHandleScope` has size 4. I think methods should require some kind of `impl` or `dyn` for scope argument, I'll look into it

File: src/handle_scope.rs (hunk 1 above)
Comment (bartlomieju) [reply src/handle_scope.rs]: Just copy-pasted from `HandleScope` so they have same API

File: tests/test_api.rs [hunk 5]
```diff
@@ -93,6 +93,63 @@ fn test_string() {
...
+  let mut locker = v8::Locker::new(&isolate);
+  isolate.enter();
+  v8::HandleScope::enter(&mut locker, |scope1| {
+    // After dropping EscapableHandleScope, we should be able to 
+    // read escaped values.  
+
+    let mut escapable_scope = v8::EscapableHandleScope::new(scope1);
+    let mut scope: v8::HandleScope = unsafe { std::mem::transmute_copy(&escapable_scope) };
+    let l1: Local<v8::Value> = cast(v8::Integer::new(&mut scope, -123));
+    let l1: Local<v8::Integer> = cast(escapable_scope.escape(l1));
```
Comment (ry) [inline tests/test_api.rs]: maybe don't write over the previous variable but use a different one

maybe put this all in a block to test scoping

File: tests/test_api.rs (hunk 3 above)
Comment (bartlomieju) [reply tests/test_api.rs]: Done and done

File: tests/test_api.rs (hunk 5 above)
Comment (bartlomieju) [reply tests/test_api.rs]: Done, now it looks super clean, I guess we could rewrite `HandleScope` to work the same way?

File: src/handle_scope.rs (hunk 2 above)
Comment (piscisaureus) [reply src/handle_scope.rs]: 'parent is unbound, so as far as the borrow checker is concerned the returned local can go anywhere. 
In reality it is only valid in the parent HandleScope.

**Pull request merged**